tests/bundle/Functional/ResourceAssertionsTrait.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+/**
+ * @copyright Copyright (C) Ibexa AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace Ibexa\Tests\Bundle\Rest\Functional;
+
+use JsonSchema\Validator;
+
+trait ResourceAssertionsTrait
+{
+    final protected static function assertResponseMatchesXmlSnapshot(string $content, ?string $file = null): void
+    {
+        self::assertStringMatchesSnapshot($content, 'xml', $file);
+    }
+
+    final protected static function assertResponseMatchesJsonSnapshot(string $content, ?string $file = null): void
+    {
+        self::assertStringMatchesSnapshot($content, 'json', $file);
+    }
+
+    /**
+     * @phpstan-param "xml"|"json"|null $type
+     */
+    final protected static function assertStringMatchesSnapshot(
+        string $content,
+        ?string $type = null,
+        ?string $file = null
+    ): void {
+        if ($file === null) {
+            $classInfo = new \ReflectionClass(static::class);
+            $class = substr(static::class, strrpos(static::class, '\\') + 1);
+            $classFilename = $classInfo->getFileName();
+            self::assertNotFalse($classFilename);
+            $file = dirname($classFilename) . '/' . $class . '.' . ($type ?? 'log');
+        }
+        if (!file_exists($file)) {
+            file_put_contents($file, rtrim($content, "\n") . "\n");
+        }
+
+        if ($type === 'xml') {
+            self::assertXmlStringEqualsXmlFile($file, $content);
+        } elseif ($type === 'json') {
+            self::assertJsonStringEqualsJsonFile($file, $content);
+        } else {
+            self::assertStringEqualsFile($file, $content);
+        }
```

Comment:
what's going on here? Can you avoid using slow Reflections please?

Reply:
Note that `ReflectionClass` is used only for tests here, and for this reason performance is barely an issue (as it is technically an operation that does not involve actual production code at all).

The purpose of this code is to get the current class location, so that any snapshot files generated are in the same directory. From a trait or abstract it's not really possible to achieve, as `__DIR__` is a compile-time information related to the file it is placed in.

Reply:
> Note that `ReflectionClass` is used only for tests here, and for this reason performance is barely an issue (as it is technically an operation that does not involve actual production code at all).
> 
> The purpose of this code is to get the current class location, so that any snapshot files generated are in the same directory. From a trait or abstract it's not really possible to achieve, as `__DIR__` is a compile-time information related to the file it is placed in.

Ok, I still have some questions about it, but let's sync internally on Monday.